source/agora/consensus/EnrollmentManager.d

```diff
@@ -717,6 +717,56 @@ public class EnrollmentManager
         return result;
     }
 
+    /***************************************************************************
+
+        Get validator's pre-image at a certain height from the validator set
+
+        Params:
+            enroll_key = The key for the enrollment in which the pre-image is
+                contained.
+            height = The block height of the preimage to get
+            result_image = will contain the PreimageInfo if exists
+
+        Returns:
+            true if getting pre-image is successfully processed
+
+    ***************************************************************************/
+
+    public bool getValidatorPreimage (const ref Hash enroll_key, ulong height,
+        out PreimageInfo result_image) @trusted nothrow
+    {
+        try
+        {
+            auto results = this.db.execute("SELECT preimage from validator_set" ~
+                        " WHERE key = ?", enroll_key.toString());
+            if (!results.empty && results.oneValue!(byte[]).length != 0)
+            {
+                PreimageInfo loaded_image =
+                    results.oneValue!(ubyte[]).deserializeFull!(PreimageInfo);
+
+                if (height <= loaded_image.height)
+                {
+                    result_image.enroll_key = loaded_image.enroll_key;
+                    result_image.hash = loaded_image.hash;
+                    result_image.height = height;
+                    foreach (i; 0 .. loaded_image.height - height)
+                    {
+                        result_image.hash = hashFull(result_image.hash);
+                        result_image.height++;
+                    }
+                }
```

Comment:
Yeah, if you change the API to just return the latest preimage, instead of a user-requested height, you can remove this whole block of code and remove an attack vector.

A very real attack vector in fact, because  what happens if we're at block 100,000 and I request the preimage at height 1 for a currently enrolled validator ? There is no bound on how much we hash here, so you would end up performing 99,998 hash operations.

Reply:
The code can be removed if the `height` parameter is removed.